### PR TITLE
GH-39523: [R] Don't override explicitly set NOT_CRAN=false when on dev version

### LINKED ISSUE #39523: [R][CI] lto nightly fails because s2n-tls can't be build with lto
### Describe the bug, including details regarding any error messages, version, and platform.

It seems the flag was changed so our exisitng removal no longer works https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=59395&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=3367cfe7-6aa5-5a88-c888-e6814d74f6eb&l=886
`Failing compile command: g++ -E -I/usr/local/include  -Wall -pedantic -flto -std=gnu++17 -xc++ -
`
See also #34349 

### Component(s)

C++, Continuous Integration, R

### PR DESCRIPTION
**Pull request by assignUser**
### Rationale for this change

The default linux build used in the lto job should not build with aws/gcs. A change in the build system changed this.

### What changes are included in this PR?

Revert to old behavior by not overriding explicitly set `NOT_CRAN=false`.

### Are these changes tested?

CI

### Are there any user-facing changes?

No
* Closes: #39523

### CODE REVIEW

**Comment by assignUser** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by assignUser** [comment]
It seems our azure queue is pretty long again (13) and this is the last in line. I have run the job locally using archery and it succeeds:
```
─  checking whether package ‘arrow’ can be installed ... [1755s/588s] OK (9m 48.2s)
─  used C++ compiler: ‘g++ (Debian 13.2.0-7) 13.2.0’
✔  checking C++ specification
     Not all R platforms support C++17
N  checking installed package size ...
     installed size is 104.8Mb
     sub-directories of 1Mb or more:
       libs  99.9Mb
       R      4.3Mb
✔  checking package directory
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -4848,6 +4848,9 @@ macro(build_awssdk)
     # Link time optimization is causing trouble like #34349
     string(REPLACE "-flto=auto" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
     string(REPLACE "-ffat-lto-objects" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
+    # Added by R with --enable/use-lto see #39523
+    string(REPLACE "-flto" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
+    string(REPLACE "-flto=10" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:4853]: Can we remove all `-flto*` flags?

```suggestion
    # "-flto" and "-flto=10" are added by R with --enable/use-lto see #39523
    string(REPLACE "-flto[^ ]*" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
    string(REPLACE "-ffat-lto-objects" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
```

Comment (assignUser) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Sounds like a future proof idea :)

**Comment by assignUser** [comment]
That works as well:
```
─  checking whether package ‘arrow’ can be installed ... [1688s/596s] OK (9m 55.9s)
─  used C++ compiler: ‘g++ (Debian 13.2.0-7) 13.2.0’
✔  checking C++ specification
     Not all R platforms support C++17
N  checking installed package size ...
     installed size is 104.8Mb
     sub-directories of 1Mb or more:
       libs  99.9Mb
       R      4.3Mb
```

**Comment by kou** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by assignUser** [comment]
hm, thats weird ... I'll test with main to confirm it fails locally. @kou anyidea why it didn't work?

**Comment by kou** [comment]
I could reproduce this by `R_ORG=rhub R_IMAGE=debian-gcc-devel-lto R_TAG=latest archery docker run -e NOT_CRAN=false -e INSTALL_ARGS=--use-LTO r` on local.
I'll take a look at this.

**Comment by assignUser** [comment]
For me it passes on main as well... :/  Thanks for taking a look!

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -4846,7 +4846,8 @@ macro(build_awssdk)
     set(AWS_LC_C_FLAGS ${EP_C_FLAGS})
     string(APPEND AWS_LC_C_FLAGS " -Wno-error=overlength-strings -Wno-error=pedantic")
     # Link time optimization is causing trouble like #34349
-    string(REPLACE "-flto=auto" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
+    # "-flto" and "-flto=10" are added by R with --enable/use-lto see #39523
+    string(REPLACE "-flto[^ ]*" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:4850]: Sorry. We need `REGEX`:

```suggestion
    string(REGEX REPLACE "-flto[^ ]*" "" AWS_LC_C_FLAGS "${AWS_LC_C_FLAGS}")
```

But there is another error:

```text
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘arrow’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/arrow/r/check/arrow.Rcheck/00LOCK-arrow/00new/arrow/libs/arrow.so':
  /arrow/r/check/arrow.Rcheck/00LOCK-arrow/00new/arrow/libs/arrow.so: undefined symbol: OPENSSL_free
```

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4850]: This warning may be related:

```text
g++ -std=gnu++17 -shared -Wall -pedantic -flto -fpic -L/opt/R-devel/lib/R/lib -L/usr/local/lib -o arrow.so RTasks.o altrep.o array.o array_to_vector.o arraydata.o arrowExports.o bridge.o buffer.o chunkedarray.o compression.o compute-exec.o compute.o config.o csv.o dataset.o datatype.o expression.o extension-impl.o feather.o field.o filesystem.o io.o json.o memorypool.o message.o parquet.o r_to_arrow.o recordbatch.o recordbatchreader.o recordbatchwriter.o safe-call-into-r-impl.o scalar.o schema.o symbols.o table.o threadpool.o type_infer.o -L/arrow/r/check/arrow.Rcheck/00_pkg_src/arrow/libarrow/arrow-14.0.1.100000264/lib -L/usr/lib/lib/x86_64-linux-gnu -L/usr/lib/lib -larrow_dataset -lparquet -larrow_acero -larrow -larrow_bundled_dependencies /usr/lib/x86_64-linux-gnu/libbz2.so -lbrotlidec -lbrotlienc -lz -lzstd -lcurl -lssl -lcrypto -L/opt/R-devel/lib/R/lib -lR
lto-wrapper: warning: Extra option to ‘-Xassembler’: --noexecstack, dropping all ‘-Xassembler’ and ‘-Wa’ options.
/tmp/RtmpoBLPp7/working_dir/RtmpRJJbvI/file77c6759da9b/aws_lc_ep-install/include/openssl/err.h:176: warning: type of ‘ERR_get_error’ does not match original declaration [-Wlto-type-mismatch]
/usr/include/openssl/err.h:417: note: return value type mismatch
  417 | unsigned long ERR_get_error(void);
      | 
/usr/include/openssl/err.h:417: note: type ‘long unsigned int’ should match type ‘uint32_t’
/usr/include/openssl/err.h:417: note: ‘ERR_get_error’ was previously declared here
/usr/include/openssl/err.h:417: note: code may be misoptimized unless ‘-fno-strict-aliasing’ is used
lto-wrapper: warning: using serial compilation of 128 LTRANS jobs
lto-wrapper: note: see the ‘-flto’ option documentation for more information
```

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4850]: OpenSSL 1 and 3 are mixed?

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4850]: Hmm, `OPENSSL_free()` is a macro both in OpenSSL 1.1 and 3.
Is a wrong header file used?

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4850]: aws-lc's `openssl/mem.h` defines `OPENSSL_free()` as a function...

**Comment by assignUser** [comment]
Yeah, sorry this is my fault, in the build refactor I also set not_cran=true if we are on a dev version and it overrides explicitly set envvar. So the issues is that this was never supposed to build like this. I will revert these changes and fix the issue in `nixlibs.R`.

**Comment by assignUser** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

File: r/tools/nixlibs.R [hunk 3]
```diff
@@ -860,7 +860,8 @@ if (is_release) {
   VERSION <- VERSION[1, 1:3]
   arrow_repo <- paste0(getOption("arrow.repo", sprintf("https://apache.jfrog.io/artifactory/arrow/r/%s", VERSION)), "/libarrow/")
 } else {
-  not_cran <- TRUE
+  # Don't override explictily set NOT_CRAN env var, as it is used in CI.
+  not_cran <- TRUE && !env_is("NOT_CRAN", "false")
```
Comment (kou) [inline r/tools/nixlibs.R:864]:
```suggestion
  not_cran <- !env_is("NOT_CRAN", "false")
```

Comment (kou) [reply r/tools/nixlibs.R:864]: The comment of this should be updated too?

Comment (assignUser) [reply r/tools/nixlibs.R:864]: > The comment of this should be updated too?

I don't understand? I already added a comment?

Comment (kou) [reply r/tools/nixlibs.R]: Ah, we have `not_cran <- env_is("NOT_CRAN", "true")` in the above line.

We may be able to simplify this like the following:

```diff
diff --git a/r/tools/nixlibs.R b/r/tools/nixlibs.R
index fe8de284b1..e89b814512 100644
--- a/r/tools/nixlibs.R
+++ b/r/tools/nixlibs.R
@@ -854,13 +854,12 @@ on_windows <- tolower(Sys.info()[["sysname"]]) == "windows"
 # For local debugging, set ARROW_R_DEV=TRUE to make this script print more
 quietly <- !env_is("ARROW_R_DEV", "true")
 
-not_cran <- env_is("NOT_CRAN", "true")
+not_cran <- !env_is("NOT_CRAN", "false")
 
 if (is_release) {
   VERSION <- VERSION[1, 1:3]
   arrow_repo <- paste0(getOption("arrow.repo", sprintf("https://apache.jfrog.io/artifactory/arrow/r/%s", VERSION)), "/libarrow/")
 } else {
-  not_cran <- TRUE
   arrow_repo <- paste0(getOption("arrow.dev_repo", "https://nightlies.apache.org/arrow/r"), "/libarrow/")
 }
 
```

Comment (assignUser) [reply r/tools/nixlibs.R]: No, this is intentional. We only want to default to setting `not_cran <- TRUE` if we are on a dev version. For release versions people have to explicitly opt in, this is needed for CRAN (as the name implies^^).

Comment (kou) [reply r/tools/nixlibs.R]: I see!

**Review by kou — APPROVED** [review]
+1

**Pull request merged**